moveit_core/planning_scene/src/urdf2collisionobject.cpp

```diff
@@ -0,0 +1,331 @@
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Rafael A. Rojas*/
+
+#include <algorithm>                      // iter_swap
+#include <eigen_conversions/eigen_msg.h>  // poseMsgToEigen
+#include <kdl_parser/kdl_parser.hpp>
+#include <memory>
+#include <moveit/collision_detection_bullet/bullet_integration/ros_bullet_utils.h>  // urdfPose2Eigen
+#include <moveit_msgs/CollisionObject.h>
+#include <robot_state_publisher/robot_state_publisher.h>
+#include <rosparam_shortcuts/rosparam_shortcuts.h>
+#include <tf2_eigen/tf2_eigen.h>  // eigenToTransform
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2_kdl/tf2_kdl.h>
+#include <tf2_ros/buffer.h>
+#include <urdf_model/link.h>
+#include <xmlrpcpp/XmlRpcValue.h>
+
+#include <geometric_shapes/mesh_operations.h>
+#include <geometric_shapes/shape_operations.h>
+#include <geometric_shapes/shape_extents.h>
+
+#include <kdl/tree.hpp>
+#include <urdf/model.h>
+
+#include <map>
+
+#include <utility>
+
+namespace planning_scene
+{
+/* Auxiliar function, used to compute the direct kinematics
+ * Taken from
+ * https://github.com/ros/robot_state_publisher/blob/ff29b5cdd0c8c1d69c4203de811120ec593e0b2f/src/robot_state_publisher.cpp#L200
+ * */
+void addChildren(const KDL::SegmentMap::const_iterator _segment,
+                 std::map<std::string, robot_state_publisher::SegmentPair>& _segments_fixed);
+
+/* Auxiliar function, used to fix names */
+std::string stripSlash(const std::string& in)
+{
+  if (not in.empty() && in[0] == '/')
+  {
+    return in.substr(1);
+  }
+  return in;
+}
+
+moveit_msgs::CollisionObject urdf_to_collision_object(const urdf::Model& _urdf_model)
+{
+  KDL::Tree kdl_tree;
+  moveit_msgs::CollisionObject result;
+
+  moveit_msgs::CollisionObject result_empty;  // empty collision object should be const
+  // ----------------------------------
+  // -------  Part I ----------
```

Comment:
Please split your huge function into small blocks, for example along the parts you already identified.